templates/invoice.json

```diff
@@ -99,116 +100,110 @@
                 ]
               ],
               [
-                "if",
-                [
-                  "is-not-null",
-                  "_S.customer.id"
-                ],
-                [
-                  [
-                    "add",
-                    "customerId",
-                    "_S.customer.id"
-                  ]
-                ]
+                "add-if",
+                "customerId",
+                "_S.customer.id"
               ],
               [
-                "if",
-                [
-                  "is-not-null",
-                  "_S.expenseAccount.id"
-                ],
-                [
-                  [
-                    "add",
-                    "expenseAccountId",
-                    "_S.expenseAccount.id"
-                  ]
-                ]
+                "add-if",
+                "expenseAccountId",
+                "_S.expenseAccount.id"
+              ],
+              [
+                "add-if",
+                "incomeAccountId",
+                "_S.incomeAccount.id"
+              ],
+              [
+                "add-if",
+                "productId",
+                "_S.product.id"
+              ],
+              [
+                "add-if",
+                "currency",
+                "_S.currency.code"
               ],
               [
                 "if",
                 [
-                  "is-not-null",
-                  "_S.incomeAccount.id"
+                  "is-not-empty",
+                  [
+                    "filter",
+                    [
+                      "is-not-null",
+                      "_.product.id"
+                    ],
+                    "_S.items"
+                  ]
                 ],
                 [
+                  "merge",
                   [
-                    "add",
-                    "incomeAccountId",
-                    "_S.incomeAccount.id"
+                    "dict",
+                    "items",
+                    [
+                      "map",
+                      [
+                        "apply",
+                        "items",
+                        "_."
+                      ],
+                      [
+                        "filter",
+                        [
+                          "is-not-null",
+                          "_.product.id"
+                        ],
+                        "_S.items"
+                      ]
+                    ]
                   ]
                 ]
-              ],
+              ]
+            ],
+            "items": [
               [
                 "if",
```

Comment:
don't need this if-test since we're checking for this in the prevous map function